scripts/openwebnet-discovery.nse

```diff
@@ -0,0 +1,246 @@
+--  |   MAC Address: 0:3:50:1:d3:11
+--  |   Device Type: F453AV
+--  |   Firmware Version: 3.0.14
+--  |   Uptime: 12d9h42m1s
+--  |   Kernel Version: 2.3.8
+--  |   Distribution Version: 3.0.1
+--  |   Date: 02.07.2017
+--  |   Time: 02:11:58
+--  |   Scenarios: 0
+--  |   Lighting: 115
+--  |   Automation: 15
+--  |   Power Management: 0
+--  |   Heating: 0
+--  |   Burglar Alarm: 12
+--  |_  Door Entry System: 0
+
+author = "Rewanth Cool"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery", "safe"}
+
+portrule = shortport.port_or_service(20000, "openwebnet")
+
+local device = {
+  [2] = "MHServer",
+  [4] = "MH200",
+  [6] = "F452",
+  [7] = "F452V",
+  [11] = "MHServer2",
+  [12] = "F453AV",
+  [13] = "H4684",
+  [15] = "F427 (Gateway Open-KNX)",
+  [16] = "F453",
+  [23] = "H4684",
+  [27] = "L4686SDK",
+  [44] = "MH200N",
+  [51] = "F454",
+  [200] = "F454 (new?)"
+}
+
+local who = {
+  [0] = "Scenarios",
+  [1] = "Lighting",
+  [2] = "Automation",
+  [3] = "Power Management",
+  [4] = "Heating",
+  [5] = "Burglar Alarm",
+  [6] = "Door Entry System",
+  [7] = "Multimedia",
+  [9] = "Auxiliary",
+  [13] = "Device Communication",
+  [14] = "Light+shutters actuators lock",
+  [15] = "CEN",
+  [16] = "Sound System",
+  [17] = "Scenario Programming",
+  [18] = "Energy Management",
+  [24] = "Lighting Management",
+  [25] = "CEN plus",
+  [1000] = "Diagnostic",
+  [1001] = "Automation Diagnostic",
+  [1004] = "Heating Diagnostic",
+  [1008] = "Door Entry System Diagnostic",
+  [1013] = "Device Diagnostic"
+}
+
+local device_dimension = {
+  ["Time"] = "0",
+  ["Date"] = "1",
+  ["IP Address"] = "10",
+  ["Net Mask"] = "11",
+  ["MAC Address"] = "12",
+  ["Device Type"] = "15",
+  ["Firmware Version"] = "16",
+  ["Hardware Version"] = "17",
+  ["Uptime"] = "19",
+  ["Micro Version"] = "20",
+  ["Date and Time"] = "22",
+  ["Kernel Version"] = "23",
+  ["Distribution Version"] = "24",
+  ["Gateway IP address"] = "50",
+  ["DNS IP address 1"] = "51",
+  ["DNS IP address 2"] = "52"
+}
+
+local ACK = "*#*1##"
+local NACK = "*#*0##"
+
+-- Initiates a socket connection
+-- Returns the socket and error message
+local function get_socket(host, port, request)
+
+  local sd, response, early_resp = comm.opencon(host, port, request)
+
+  if sd == nil then
+    return nil, "Socket connection error."
+  end
+
+  if not response then
+    return nil, "Poor internet connection or no response."
+  end
+
+  if response == NACK then
+    return nil, "Received a negative ACK as response."
+  end
+
+  return sd, nil
+end
+
+local function get_response(sd, request)
+
+  local res = {}
+  local status, data
+
+  sd:send(request)
+
+  repeat
+    status, data = sd:receive_buf("##", true)
+    if status and data ~= ACK then
+      table.insert(res, data)
+    end
+    if data == ACK then
+      break
+    end
+
+    -- If response is NACK, it means the request method is not supported
+    if data == NACK then
```

Comment:
And yet another tricky one: timeouts. If the timeout expires, we may end up reading a response to one probe as the answer to the next probe. There are 2 approaches we must take to avoid this:

1. Increase the [`request_timeout`](https://nmap.org/nsedoc/lib/comm.html) when connecting with `comm.opencon`, but do not increase the `timeout` or `connect_timeout`. Requests may take longer because the gateway has to contact the slave devices. Probably use 10 seconds (my test system took 8.2 seconds, greater than the connect_timeout (1.4s) plus the request_timeout (6s) by default).
2. To the extent possible, we should try to check the response and synchronize. Here was the output of my mixup:
    ```
    NSE: [openwebnet-discovery 78.216.11.170:20000] Fetching the list of Heating devices.
    NSE: TCP 192.168.1.23:46051 > 78.216.11.170:20000 | *#4*0##
    NSOCK INFO [11.4180s] nsock_write(): Write request for 7 bytes to IOD #1 EID 451 [78.216.11.170:20000]
    NSOCK INFO [11.4180s] nsock_trace_handler_callback(): Callback: WRITE SUCCESS for EID 451 [78.216.11.170:20000]
    NSE: TCP 192.168.1.23:46051 > 78.216.11.170:20000 | SEND
    NSOCK INFO [11.4180s] nsock_read(): Read request from IOD #1 [78.216.11.170:20000] (timeout: 7434ms) EID 458
    NSOCK INFO [18.8520s] nsock_trace_handler_callback(): Callback: READ TIMEOUT for EID 458 [78.216.11.170:20000]
    NSE: [openwebnet-discovery 78.216.11.170:20000] Fetching the list of Burglar Alarm devices.
    NSE: TCP 192.168.1.23:46051 > 78.216.11.170:20000 | *#5*0##
    NSOCK INFO [18.8530s] nsock_write(): Write request for 7 bytes to IOD #1 EID 467 [78.216.11.170:20000]
    NSOCK INFO [18.8530s] nsock_trace_handler_callback(): Callback: WRITE SUCCESS for EID 467 [78.216.11.170:20000]
    NSE: TCP 192.168.1.23:46051 > 78.216.11.170:20000 | SEND
    NSOCK INFO [18.8530s] nsock_read(): Read request from IOD #1 [78.216.11.170:20000] (timeout: 7434ms) EID 474
    NSOCK INFO [19.6130s] nsock_trace_handler_callback(): Callback: READ SUCCESS for EID 474 [78.216.11.170:20000] (6 bytes): *#*0##
    NSE: TCP 192.168.1.23:46051 < 78.216.11.170:20000 | *#*0##
    NSE: [openwebnet-discovery 78.216.11.170:20000] Fetching the list of Door Entry System devices.
    NSE: TCP 192.168.1.23:46051 > 78.216.11.170:20000 | *#6*0##
    NSOCK INFO [19.6130s] nsock_write(): Write request for 7 bytes to IOD #1 EID 483 [78.216.11.170:20000]
    NSOCK INFO [19.6130s] nsock_trace_handler_callback(): Callback: WRITE SUCCESS for EID 483 [78.216.11.170:20000]
    NSE: TCP 192.168.1.23:46051 > 78.216.11.170:20000 | SEND
    NSOCK INFO [19.6140s] nsock_read(): Read request from IOD #1 [78.216.11.170:20000] (timeout: 7434ms) EID 490
    NSOCK INFO [23.1590s] nsock_trace_handler_callback(): Callback: READ SUCCESS for EID 490 [78.216.11.170:20000] (7 bytes): *5*0*##
    NSE: TCP 192.168.1.23:46051 < 78.216.11.170:20000 | *5*0*##
    ```
    So Heating (4) timed out, but the server sent a NACK which we interpreted as the answer for Burglar Alarm (5). Then we sent Door Entry (6) and got a response for 5. We should detect that the response doesn't match the request. Maybe extract the WHO value from the response and use that instead of just assuming it matches (only works for non-NACK responses). That would at least avoid mixing up answers, even if it means we accidentally drop some. Better to not report something than to report it incorrectly.